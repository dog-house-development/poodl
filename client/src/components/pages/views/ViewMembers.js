import React, { Component } from 'react';
import { connect } from 'react-redux';
import _ from 'lodash';

import { fetchMembers } from '../../../actions/memberActions';
import DataGrid from '../../ui/DataGrid';

export class ViewMembers extends Component {
    componentDidMount() {
        // call redux action to retrieve all members from api
        this.props.getMembers();
    }

    getDataGridContent() {
        // choose what we want to display out of the members data
        let data = [];
        _.each(this.props.members, member => {
            // we want members' names and emails
            data.push({ firstName: member.firstName, email: member.email, key: member._id });
        });
        return data;
    }

    render() {
        return (
            <div className="view-all-container">
                <h1>View All Members</h1>
                <DataGrid data={this.getDataGridContent()} loading={this.props.loading} />
            </div>
        );
    }
}

export const mapStateToProps = (state, props) => {
    return {
        loading: state.members.loading,
        members: state.members.all.data,
        errors: state.errors
    };
};

export const mapDispatchToProps = dispatch => {
    return {
        getMembers: () => dispatch(fetchMembers())
    };
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(ViewMembers);
