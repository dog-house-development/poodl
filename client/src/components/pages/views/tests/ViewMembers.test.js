import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import _ from 'lodash';

import { ViewMembers, mapStateToProps, mapDispatchToProps } from '../ViewMembers';

configure({ adapter: new Adapter() });

describe('ViewMembers tests', () => {
    let state, props, wrapper, instance;
    const setInstanceAndWrapper = (_props = {}, _state = {}) => {
        state = _.assign(
            {},
            {
                members: {
                    loading: false,
                    all: {
                        data: [
                            {
                                _id: '123',
                                firstName: 'Big',
                                lastName: 'Tup',
                                membershipDate: '02/20/2020',
                                email: 'bigtup@nowhere.com'
                            },
                            {
                                _id: '321',
                                firstName: 'Lil',
                                lastName: 'Tup',
                                membershipDate: '02/02/2002',
                                email: 'liltup@nowhere.com'
                            }
                        ]
                    }
                },
                errors: {}
            },
            _state
        );
        props = _.assign({}, _props);
        wrapper = shallow(
            <ViewMembers
                {..._.assign(
                    {},
                    props,
                    mapStateToProps(state, props),
                    mapDispatchToProps(jasmine.createSpy('dispatch'))
                )}
            />
        );
        instance = wrapper.instance();
    };

    beforeEach(() => {
        setInstanceAndWrapper();
    });

    describe('mapStateToProps', () => {
        it('should map state to props', () => {
            expect(mapStateToProps(state, props)).toEqual({
                members: [
                    {
                        _id: '123',
                        firstName: 'Big',
                        lastName: 'Tup',
                        membershipDate: '02/20/2020',
                        email: 'bigtup@nowhere.com'
                    },
                    {
                        _id: '321',
                        firstName: 'Lil',
                        lastName: 'Tup',
                        membershipDate: '02/02/2002',
                        email: 'liltup@nowhere.com'
                    }
                ],
                errors: {},
                loading: false
            });
        });
    });

    describe('mapDispatchToProps', () => {
        it('should map dispatch to props', () => {
            const dispatch = jest.fn();
            expect(JSON.stringify(mapDispatchToProps(dispatch))).toEqual(JSON.stringify({ getMembers: () => {} }));
        });
    });

    describe('componentDidMount', () => {
        it('should run without breaking', () => {
            spyOn(instance, 'componentDidMount');
            instance.componentDidMount();
        });
    });

    describe('getDataGridContent', () => {
        it('should return filtered members data', () => {
            expect(instance.getDataGridContent()).toEqual([
                {
                    key: '123',
                    firstName: 'Big',
                    lastName: 'Tup',
                    membershipDate: '02/20/2020',
                    email: 'bigtup@nowhere.com'
                },
                {
                    key: '321',
                    firstName: 'Lil',
                    lastName: 'Tup',
                    membershipDate: '02/02/2002',
                    email: 'liltup@nowhere.com'
                }
            ]);
        });
    });

    describe('render', () => {
        it('should render correctly', () => {
            expect(wrapper).toMatchSnapshot();
        });
    });
});
