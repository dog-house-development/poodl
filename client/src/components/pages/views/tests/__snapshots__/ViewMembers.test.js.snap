// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewMembers tests render should render correctly 1`] = `
<div
  className="view-all-container"
>
  <h1>
    View All Members
  </h1>
  <DataGrid
    data={
      Array [
        Object {
          "email": "bigtup@nowhere.com",
          "firstName": "Big",
          "key": "123",
          "lastName": "Tup",
          "membershipDate": "02/20/2020",
        },
        Object {
          "email": "liltup@nowhere.com",
          "firstName": "Lil",
          "key": "321",
          "lastName": "Tup",
          "membershipDate": "02/02/2002",
        },
      ]
    }
    loading={false}
  />
</div>
`;
